### DISCUSSION TITLE
local builds with x86_64h

### DISCUSSION THREAD

**Discussion by timblechmann** [Homebrew/discussions#1266] [Everyday usage]
the intel packages of homebrew are compiled as `x86_64`. however i wonder if there's any way to rebuild packages (potentially just locally) with `-arch x86_64h` to enable avx codegen?

**Comment by danielnachun** [comment]
I believe Homebrew specifies `-march=native` when you build locally (I know it does on Linux by default).  If you compile locally on a machine whose CPU supports AVX512 instructions, I think they should be enabled with no further intervention required.  To build a formula locally you'd just use `brew install -s FORMULA` so that it builds from source even if a bottle is available.

**Reply by timblechmann** [reply]
interesting. i've just tried, but it seems that clang is invoked with `-march=nehalem`, which does not enable avx. but is probably related to the minimum deployment target

**Reply by danielnachun** [reply]
Right, in CI we always try pick a conservative minimum deployment target so the binaries work as widely as possible, but I guess on macOS (with `clang` at least, no idea for GCC), we are defaulting to `nehalem` in your case even when building locally.  

On Linux, where I think this is a more common issue due the range of hardware being used, the `HOMEBREW_ARCH` environment variable can be used to set the level of optimization.  But in https://github.com/Homebrew/brew/blob/7a9e7f92042d9d15d5824020b161266bf3fceb2e/Library/Homebrew/env_config.rb, it clearly indicates that is a Linux-only variable.

@fxcoudert, @mistydemeo or @claui may have more knowledge on how to set optimization levels in `clang` on macOS as it's clearly handled somewhat differently from Linux.  

If I might ask, which formula(e) are you trying to build with AVX512 support?  I've got a PR open right now for embrue https://github.com/Homebrew/homebrew-core/pull/74332, and in that particular formula we're trying to figure out how to handle AVX512 support, among other issues.  So what we learn here could be useful in figuring that out too.

**Reply by mistydemeo** [reply]
We've currently disabled building bottles with AVX because they won't run under Rosetta 2. For the period where that's a supported configuration for us, we don't want to deliver bottles some users can't run. You can see the comment here: https://github.com/Homebrew/brew/blob/master/Library/Homebrew/extend/os/mac/hardware.rb

If building locally from source, and not via `--build-bottle`, we use `--march=native` instead of the compatible option. The logic is here: https://github.com/Homebrew/brew/blob/bf7fe45e8998e56e6690347a0192c454b8cb203b/Library/Homebrew/extend/os/linux/extend/ENV/shared.rb

In general, on macOS, we specify the CPU based on hardware support. We use the minimum requirements for the OS we're building for and build for that architecture.

**Reply by danielnachun** [reply]
Ah so it's not actually that different from Linux after all.  I wonder why @timblechmann only gets `--march=nehalem` when building locally from source?  I guess we'll have to see what formula or formulae are being built to see if there are other factors at play.

I did not know Rosetta 2 doesn't support AVX but I see now that Apple very clearly states that AVX, AVX2, and AVX512 aren't supported.  There are some formulae like `embree` and `openblas` which seem to use runtime checks for ISA support so that in theory one could compile them with support up through AVX512, and then at runtime the package knows not to use unsupported instructions.  It could be interesting to explore how to handle those "edge cases" though they would need quite a bit of testing to fully understand.

**Reply by timblechmann** [reply]
> We've currently disabled building bottles with AVX because they won't run under Rosetta 2

not being familiar how bottles are built, have there been any thoughts on providing UBs for `x86_64` and `x86_64h`? i suppose rosetta would then fall back to the `x86_64` code path. of course it doubles compile times/binary sizes and requires some lipoing

> In general, on macOS, we specify the CPU based on hardware support. We use the minimum requirements for the OS we're building for and build for that architecture.

i'm wondering if this is actually necessary: `-mmacosx-​version-min` implicitly enables all instruction sets that are available for the specific version (iirc this is mainly sse4.2 being enabled from one macos version onwards (iirc 10.13)
